File: execution/runner.py
# execution/runner.py (Modified for Clean Output)
import sys
import json
import importlib.util
from copy import deepcopy
import os
import traceback

CODE_FILE = '/sandbox/user_code.py'
INPUT_FILE = '/sandbox/input.json'

# Helper to log consistently to stderr
def log_stderr(message):
    print(f"RunnerLog: {message}", file=sys.stderr)

def run_code():
    try:
        # --- Read Input Data ---
        log_stderr("Script started.")
        if not os.path.exists(INPUT_FILE):
            raise FileNotFoundError(f"Input file not found at {INPUT_FILE}")

        with open(INPUT_FILE, 'r', encoding='utf-8') as f: # Specify encoding
            input_raw = f.read()
            log_stderr(f"Read input data: {input_raw}")
            input_data = json.loads(input_raw)

        function_name = input_data.get('functionName')
        input_args = input_data.get('inputArgs')

        if not function_name or not isinstance(input_args, list):
             raise ValueError('Invalid input data format.')
        log_stderr(f"Executing function: {function_name} with args: {input_args!r}") # Use repr for args

        # --- Load User Code ---
        if not os.path.exists(CODE_FILE):
             raise FileNotFoundError(f"Code file not found at {CODE_FILE}")

        module_name = "user_module"
        spec = importlib.util.spec_from_file_location(module_name, os.path.abspath(CODE_FILE))
        if spec is None or spec.loader is None:
             raise ImportError(f"Could not create module spec from code file: {CODE_FILE}")

        user_module = importlib.util.module_from_spec(spec)
        sys.modules[module_name] = user_module
        log_stderr("Loading user module...")
        spec.loader.exec_module(user_module)
        log_stderr("User module loaded.")

        if not hasattr(user_module, function_name):
             raise AttributeError(f"Function '{function_name}' not found in user code.")

        user_func = getattr(user_module, function_name)

        # --- Execute User Code ---
        result = None
        args_copy = deepcopy(input_args)
        log_stderr("Executing user function...")

        # Clear stdout buffer before calling user code (just in case)
        sys.stdout.flush()

        if function_name == "reverseString":
             user_func(args_copy[0]) # Modifies args_copy[0] in-place
             result = args_copy[0]   # Result is the modified list
             log_stderr("Executed reverseString (in-place).")
        else:
             result = user_func(*args_copy) # Standard call
             log_stderr("Executed standard function call.")

        # --- Prepare and Output Result ---
        final_output = result
        log_stderr(f"Execution successful. Raw Result: {final_output!r}")

        # --- Write ONLY JSON SUCCESS to STDOUT ---
        success_payload = {"status": "success", "output": final_output}
        success_json = json.dumps(success_payload, ensure_ascii=False)
        # Print ONLY this JSON string to stdout
        print(success_json, file=sys.stdout, flush=True) # Add flush=True here too
        # --- End STDOUT write ---

        log_stderr("Successfully wrote JSON to stdout.")

    except Exception as e:
        # --- Handle Errors ---
        error_type = type(e).__name__
        error_message = str(e)
        log_stderr(f"!!! Runner script error: {error_type}: {error_message}")
        tb_str = traceback.format_exc()
        log_stderr(f"Traceback:\n{tb_str}")

        # --- Write ONLY JSON ERROR to STDERR ---
        error_payload = {
            "status": "error",
            "type": error_type,
            "message": error_message
            # Avoid sending traceback to frontend
        }
        error_json = json.dumps(error_payload, ensure_ascii=False)
        print(error_json, file=sys.stderr, flush=True) # Print error JSON to stderr and flush
        # --- End STDERR write ---
        sys.exit(1) # Exit with non-zero code

    log_stderr("Script finished normally.")
    # No need for extra flushes here if print has flush=True

if __name__ == "__main__":
    run_code()

File: execution/runner.js
// execution/runner.js (Corrected Paths and Output Handling)
const fs = require('fs');
const path = require('path');

// --- Correct Paths relative to container structure ---
const SANDBOX_DIR = '/sandbox'; // Base directory where temp files are mounted
const CODE_FILE = path.join(SANDBOX_DIR, 'user_code.js'); // Correct path to user code
const INPUT_FILE = path.join(SANDBOX_DIR, 'input.json');  // Correct path to input data
// ----------------------------------------------------

// Helper to log consistently to stderr *without* extra characters
function logError(...args) {
    // Use console.error which writes to stderr by default
    console.error("RunnerLog:", ...args);
}

logError("Node runner script started."); // Keep only one start message

try {
    logError(`Runner Step 1: Checking input file: ${INPUT_FILE}`);
    if (!fs.existsSync(INPUT_FILE)) {
        throw new Error(`Input file not found: ${INPUT_FILE}`);
    }
    const inputRaw = fs.readFileSync(INPUT_FILE, 'utf8');
    logError("Runner Step 2: Input file read. Parsing JSON...");
    const inputData = JSON.parse(inputRaw);
    // Use JSON.stringify for logging objects cleanly
    logError("Runner Step 3: JSON Parsed:", JSON.stringify(inputData));

    const { functionName, inputArgs } = inputData;
    if (!functionName || !Array.isArray(inputArgs)) {
        throw new Error('Invalid input data format (missing functionName or inputArgs is not array).');
    }

    logError(`Runner Step 4: Checking code file: ${CODE_FILE}`);
    if (!fs.existsSync(CODE_FILE)) {
        throw new Error(`Code file not found: ${CODE_FILE}`);
    }

    logError(`Runner Step 5: Requiring code module...`);
    // Resolve the absolute path just in case require needs it
    const userCodeModule = require(path.resolve(CODE_FILE));
    logError("Runner Step 6: Code module required. Checking function...");

    if (typeof userCodeModule[functionName] !== 'function') {
        throw new Error(`Function '${functionName}' not found or not exported in user code: ${CODE_FILE}`);
    }
    const userFunction = userCodeModule[functionName];
    logError("Runner Step 7: Executing user function...");

    let result;
    // Execute user code
    // Special handling for reverseString if needed (though less common in JS)
    // if (functionName === "reverseString") { ... } else { ... }
    // Assuming standard call pattern for now:
    result = userFunction(...inputArgs);
    logError("Runner Step 8: Function executed.");


    const finalOutput = result; // JS handles undefined/null okay in JSON stringify
    logError(`Runner Step 9: Execution finished. Raw Result:`, finalOutput); // Log raw result

    // --- Write ONLY JSON SUCCESS to STDOUT ---
    const successPayload = { status: 'success', output: finalOutput };
    // Use process.stdout.write directly to avoid potential console.log formatting issues
    process.stdout.write(JSON.stringify(successPayload)); // NO newline added by default
    // --- End STDOUT write ---

    logError("Runner Step 10: Wrote success output to stdout.");

} catch (error) {
    logError("Runner Step E1: Error caught.", error.message); // Log the error message
    // Log stack trace for better debugging if needed
    logError("Stack Trace:", error.stack);

    // --- Write ONLY JSON ERROR to STDERR ---
    const errorPayload = {
        status: 'error',
        type: error.name || 'Error', // Get error type (e.g., 'Error', 'TypeError')
        error: error.message       // Get error message string
    };
    // Use process.stderr.write directly to avoid potential console.error formatting issues
    process.stderr.write(JSON.stringify(errorPayload) + '\n'); // Add newline for clarity in logs if read line-by-line
    // --- End STDERR write ---

    logError("Runner Step E2: Wrote error JSON to stderr.");
    process.exit(1); // Crucial: Exit with a non-zero code to indicate failure
}

logError("Runner finished normally."); // Should only be reached on success
// process.exit(0); // Optional: Explicit success exit code